transport/nats/publisher.go

```diff
@@ -0,0 +1,138 @@
+package nats
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/go-kit/kit/endpoint"
+	"github.com/nats-io/go-nats"
+	"time"
+)
+
+// Publisher wraps a URL and provides a method that implements endpoint.Endpoint.
+type Publisher struct {
+	publisher *nats.Conn
+	subject   string
+	reply     string
+	enc       EncodeRequestFunc
+	dec       DecodeResponseFunc
+	before    []RequestFunc
+	after     []ClientResponseFunc
+	async     bool
+	timeout   time.Duration
+}
+
+// NewClient constructs a usable Publisher for a single remote method.
+func NewPublisher(
+	publisher *nats.Conn,
+	subject string,
+	reply string,
+	enc EncodeRequestFunc,
+	dec DecodeResponseFunc,
+	options ...PublisherOption,
+) *Publisher {
+	c := &Publisher{
+		publisher: publisher,
+		subject:   subject,
+		reply:     reply,
+		enc:     enc,
+		dec:     dec,
+		before:  []RequestFunc{},
+		after:   []ClientResponseFunc{},
+		async:   false,
+		timeout: 10 * time.Second,
+	}
+	for _, option := range options {
+		option(c)
+	}
+	return c
+}
+
+// PublisherOption sets an optional parameter for clients.
+type PublisherOption func(*Publisher)
+
+// SetPublisher sets the underlying HTTP client used for requests.
+// By default, http.DefaultClient is used.
+func SetPublisher(publisher *nats.Conn) PublisherOption {
+	return func(c *Publisher) { c.publisher = publisher }
+}
+
+// PublisherBefore sets the RequestFuncs that are applied to the outgoing HTTP
+// request before it's invoked.
+func PublisherBefore(before ...RequestFunc) PublisherOption {
+	return func(c *Publisher) { c.before = append(c.before, before...) }
+}
+
+// PublisherAfter sets the ClientResponseFuncs applied to the incoming HTTP
+// request prior to it being decoded. This is useful for obtaining anything off
+// of the response and adding onto the context prior to decoding.
+func PublisherAfter(after ...ClientResponseFunc) PublisherOption {
+	return func(c *Publisher) { c.after = append(c.after, after...) }
+}
+
+// Endpoint returns a usable endpoint that invokes the remote endpoint.
+func (p Publisher) Endpoint() endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		msg := nats.Msg{Subject: p.subject, Reply: p.reply}
+
+		if err := p.enc(ctx, &msg, request); err != nil {
+			return nil, err
+		}
+
+		for _, f := range p.before {
+			ctx = f(ctx, &msg)
+		}
+
+		if !p.async {
+			resp, err := p.publisher.Request(p.subject, msg.Data, p.timeout)
```

Comment:
Since you already have a context you can use `RequestWithContext` instead and remove the timeout field.